### PR TITLE
Tofino target changes

### PR DESCRIPTION
**Pull request by ravi861**
These changes still would not build without an updated bfsde.BUILD or an alternative.

### CODE REVIEW

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 1]
```diff
@@ -522,10 +524,17 @@ ::util::Status GetAllEntries(
   tdi::DevMgr::getInstance().deviceGet(0, &device);
   tdi::Flags *flags = new tdi::Flags(0);
   uint32 entries = 0;
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:526]: I would suggest adding the following at the top of the file:
```
#ifdef DPDK_TARGET
  typedef tdi_rt_table_type_e sde_table_type;
  #define SDE_TABLE_TYPE_COUNTER TDI_RT_TABLE_TYPE_COUNTER
  #define SDE_TABLE_TYPE_METER   TDI_RT_TABLE_TYPE_METER
#elif TOFINO_TARGET
  typedef tdi_tofino_table_type_e sde_table_type;
  #define SDE_TABLE_TYPE_COUNTER TDI_TOFINO_TABLE_TYPE_COUNTER
  #define SDE_TABLE_TYPE_METER   TDI_TOFINO_TABLE_TYPE_METER
#endif
```
and then generalizing the following code as:
```
auto table_type = static_cast<sde_table_type>(
    table->tableInfoGet()->tableTypeGet());
 if (table_type == SDE_TABLE_TYPE_COUNTER ||
    table_type == SDE_TABLE_TYPE_METER) {
```
This should make the code easier to read and maintain.

File: stratum/hal/lib/tdi/tdi_id_mapper.cc [hunk 2]
```diff
@@ -203,20 +203,33 @@ ::util::Status TdiIdMapper::BuildActionProfileMapping(
   for (const auto* table : tdi_tables) {
     std::string table_name;
     tdi_id_t table_id;
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_id_mapper.cc:205]: I would suggest adding the following at the top of the file:
```
#ifdef DPDK_TARGET
  typedef tdi_rt_table_type_e sde_table_type;
  #define SDE_TABLE_TYPE_ACTION_PROFILE TDI_RT_TABLE_TYPE_ACTION_PROFILE
  #define SDE_TABLE_TYPE_SELECTOR       TDI_RT_TABLE_TYPE_SELECTOR
#elif TOFINO_TARGET
  typedef tdi_tofino_table_type_e sde_table_type;
  #define SDE_TABLE_TYPE_ACTION_PROFILE TDI_TOFINO_TABLE_TYPE_ACTION_PROFILE
  #define SDE TABLE_TYPE_SELECTOR       TDI_TOFINO_TABLE_TYPE_SELECTOR
#endif
```
and then changing the typecast to:
```
auto table_type =
    static_cast<sde_type_type>(table->tableInfoGet()->tableTypeGet());
```
You can also remove the TODO comment.

This should make the code easier to read and maintain.

File: stratum/hal/lib/tdi/tdi_id_mapper.cc [hunk 3]
```diff
@@ -203,20 +203,33 @@ ::util::Status TdiIdMapper::BuildActionProfileMapping(
...
-    auto table_type = static_cast<tdi_rt_table_type_e>(
-	table->tableInfoGet()->tableTypeGet());
+    auto table_type =
+        static_cast<tdi_rt_table_type_e>(table->tableInfoGet()->tableTypeGet());
+#elif TOFINO_TARGET
+    auto table_type = static_cast<tdi_tofino_table_type_e>(
+        table->tableInfoGet()->tableTypeGet());
+#endif
     table_id = table->tableInfoGet()->idGet();
     table_name = table->tableInfoGet()->nameGet();
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_id_mapper.cc:217]: If you add the definitions described in the comment above, you can change the conditionals below to
```
if (table_type == SDE_TABLE_TYPE_ACTION_PROFILE)
```
and
```
if (table_type == SDE_TABLE_TYPE_SELECTOR)
```
You can also remove the TODO comment.

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 4]
```diff
@@ -1232,21 +1248,27 @@ ::util::Status TdiSdeWrapper::HotplugPort(
...
 ::util::Status TdiSdeWrapper::AddPort(int device, int port, uint64 speed_bps,
                                      FecMode fec_mode) {
+#ifdef DPDK_TARGET
   auto port_attrs = absl::make_unique<port_attributes_t>();
-#ifndef P4OVS_CHANGES
+  RETURN_IF_TDI_ERROR(bf_pal_port_add(static_cast<bf_dev_id_t>(device),
+                                      static_cast<bf_dev_port_t>(port),
+                                      port_attrs.get()));
+#elif TOFINO_TARGET
   // P4OVS_CHANGES: 'bf_' and 'Bf' are Barefoot-specific names.
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:1264]: Please delete this comment.

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 5]
```diff
@@ -1032,7 +1041,7 @@ ::util::StatusOr<int> AutonegHalToBf(TriState autoneg) {
   }
 }
 
-#ifndef P4OVS_CHANGES
+#ifdef TOFINO_TARGET
 
 // P4OVS_CHANGES: 'Bf' is a Barefoot-specific name.
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:1046]: Please delete this comment. 'Bf' is fine in Tofino-specific code.

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 6]
```diff
@@ -996,7 +1005,7 @@ bf_status_t sde_port_status_callback(bf_dev_id_t device, bf_dev_port_t dev_port,
   return status.ok() ? BF_SUCCESS : BF_INTERNAL_ERROR;
 }
 
-#ifndef P4OVS_CHANGES
+#ifdef TOFINO_TARGET
 // P4OVS_CHANGES: 'Bf' is a Barefoot-specific name.
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:1009]: Please delete this comment. 'Bf' is fine in Tofino-specific code.

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 7]
```diff
@@ -522,10 +524,17 @@ ::util::Status GetAllEntries(
...
   if (table_type == TDI_RT_TABLE_TYPE_COUNTER ||
       table_type == TDI_RT_TABLE_TYPE_METER) {
+#elif TOFINO_TARGET
+  auto table_type = static_cast<tdi_tofino_table_type_e>(
+      table->tableInfoGet()->tableTypeGet());
+  if (table_type == TDI_TOFINO_TABLE_TYPE_COUNTER ||
+      table_type == TDI_TOFINO_TABLE_TYPE_METER) {
+#endif
     size_t table_size;
 // P4OVS_CHANGES: This conditional is Tofino-specific. What should we do
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:539]: Please change comment to:
```
// TODO: 'BF' is Barefoot-specific.
```

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 8]
```diff
@@ -1097,6 +1106,7 @@ ::util::StatusOr<PortState> TdiSdeWrapper::GetPortState(int device, int port) {
 
 ::util::Status TdiSdeWrapper::GetPortCounters(int device, int port,
                                              PortCounters* counters) {
+#ifdef DPDK_TARGET
   uint64_t stats[BF_PORT_NUM_COUNTERS] = {0};
 
   // P4OVS_CHANGES: 'bf_' is a Barefoot-specific name.
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:1112]: Please change P4OVS_CHANGES to TODO.

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 9]
```diff
@@ -1256,12 +1278,16 @@ ::util::Status TdiSdeWrapper::AddPort(
   static int port_in;
   static int port_out;
 
-  auto port_attrs = absl::make_unique<port_attributes_t>();
-#ifndef P4OVS_CHANGES
+#ifdef TOFINO_TARGET
   // P4OVS_CHANGES: 'bf_' and 'Bf' are Barefoot-specific names.
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:1282]: Please delete this comment.

**Review by ffoulkes — APPROVED** [review]
With the noted changes, these look good. This is a terrific boost. Thank you.

Can this be merged into my code on `split-arch`? If you do, will your changes become part of the Draft PR? I'm still on the github learning curve.

**Comment by ffoulkes** [comment]
Actually, the question of whether the changes show up automatically in the Draft PR is academic. There's more work to be done before this gets pushed to ipdk-io. I can always cancel the Draft PR and generate a real one when the time's right.

I'm going to shift to using topic branches when I do my next update, so we can review each other's contributions.

**Comment by ravi861** [comment]
> 

It will show up in your draft PR once I merge this to split-arch. We can keep this draft PR open and keep merging from topic branches.

File: stratum/hal/lib/tdi/tdi_id_mapper.cc (hunk 2 above)
Comment (ravi861) [reply stratum/hal/lib/tdi/tdi_id_mapper.cc:215]: Done

File: stratum/hal/lib/tdi/tdi_id_mapper.cc (hunk 3 above)
Comment (ravi861) [reply stratum/hal/lib/tdi/tdi_id_mapper.cc:220]: Done

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc (hunk 1 above)
Comment (ravi861) [reply stratum/hal/lib/tdi/tdi_sde_wrapper.cc:536]: Done

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc (hunk 6 above)
Comment (ravi861) [reply stratum/hal/lib/tdi/tdi_sde_wrapper.cc]: Done

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc (hunk 5 above)
Comment (ravi861) [reply stratum/hal/lib/tdi/tdi_sde_wrapper.cc]: Done

File: stratum/hal/lib/tdi/tdi_sde_wrapper.cc [hunk 10]
```diff
@@ -522,12 +534,12 @@ ::util::Status GetAllEntries(
   tdi::DevMgr::getInstance().deviceGet(0, &device);
   tdi::Flags *flags = new tdi::Flags(0);
   uint32 entries = 0;
-  auto table_type = static_cast<tdi_rt_table_type_e>(
+  auto table_type = static_cast<sde_table_type_e>(
```
Comment (ffoulkes) [inline stratum/hal/lib/tdi/tdi_sde_wrapper.cc:537]: There are a couple of typos:
- `sde_table_type_e` should be `sde_table_type` (above)
- `SDE_RT_TABLE_TYPE_METER` should be `SDE_TABLE_TYPE_METER` (below)

**Review by ffoulkes — APPROVED** [review]
LGTM (once the typos are fixed)

**Comment by ffoulkes** [comment]
IPDK practice is that commits need to be signed (`git commit -s`)

**Comment by ffoulkes** [comment]
> These changes still would not build without an updated bfsde.BUILD or an alternative.

I propose that we create a new `tofino.BUILD` file to define the external repository, plus a `tofino.bzl` file to define the context within `tdi`, with accompanying changes to `deps.bzl` and `WORKSPACE`. The process probably involve some trial and error; I've never done this before.

**Pull request merged**